<div align="center">

# Programming for Data Analytics Project
## Residential Property Price Register Analysis
***

</div>

<p align="center">
    <img width="959" height="539" src="./images/housing_predictions.webp" alt="Sublime's custom image"/
    >
    </p>

###### [Image from Forbes.com](https://www.forbes.com/advisor/mortgages/real-estate/housing-market-predictions/)

### Table of Contents


1. About the Project
   
2. Import Libraries
   
3. Load Data

4. Data Exploration
   
    4.1  Check for the DataFrame

    4.2  Check for the DataFrame dimensionality with pandas .info() method

    4.3  Generate descriptive statistics with pandas .describe method

    4.4  Check for missing values   

5. Data Analysis
   
    5.1  Analysis of Property Prices for all years in Ireland (2010-2024)
    
     - 5.1.1    Average property prices by county

     - 5.1.2    Average property prices by month

     - 5.1.3    Number of occurrences of each property type

     - 5.1.4    Full Market price comparison

     - 5.1.5    Number of property sales by county

     - 5.1.6    County with the highest and lowest average property prices
  
    5.2 Analysis of Property Prices for all years in Mayo (2010-2024)
 
     - 5.2.1 

     - 5.2.2 

     - 5.2.3 
 
6.  References

### 1. About the Project
***

### 2. Import the Libraries
***

I imported the following libraries to plot the dataset.

- `matplotlib.pyplot`: Essential for creating static, animated, and interactive visualizations in Python. It is closely integrated with NumPy and provides a MATLAB-like interface for creating plots and visualizations.
- `numpy`: It contains functionality for multidimensional arrays, high-level mathematical functions such as linear algebra operations.
- `pandas`: Fundamental data analysis and manipulation library built on top of the Python programming language. It offers data structures and operations for manipulating numerical tables and time series.
- `seaborn`: Seaborn is a Python data visualization library based on matplotlib. It provides a high-level interface for drawing attractive and informative statistical graphics.
- `datetime`: The datetime is a built-in module that provides classes for manipulating dates and times

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import datetime

### 3. Load the Data
***
Load the dataset from the Residential Property Price Register website https://www.propertypriceregister.ie/


In [5]:
df=pd.read_csv("./data/ppr_all.csv")

C:\Users\curleys\AppData\Local\Temp\ipykernel_30620\1623581419.py:1: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv("./data/ppr_all.csv")


### 4. Data Exploration
***

#### 4.1 Check the dataframe using df.head

Lets have an initial glance at the data

In [66]:
df.head(5)

Date of Sale (dd/mm/yyyy)                                         Address  \
0                01/01/2010          5 Braemor Drive, Churchtown, Co.Dublin   
1                03/01/2010  134 Ashewood Walk, Summerhill Lane, Portlaoise   
2                04/01/2010             1 Meadow Avenue, Dundrum, Dublin 14   
3                04/01/2010                         1 The Haven, Mornington   
4                04/01/2010                   11 Melville Heights, Kilkenny   

     County Eircode  Price in Euro Not Full Market Price VAT Exclusive  \
0    Dublin     NaN       343000.0                    No            No   
1     Laois     NaN       185000.0                    No           Yes   
2    Dublin     NaN       438500.0                    No            No   
3     Meath     NaN       400000.0                    No            No   
4  Kilkenny     NaN       160000.0                    No            No   

                 Description of Property  \
0  Second-Hand Dwelling house /Apartment   
1          New Dwelling house /Apartment   
2  Second-Hand Dwelling house /Apartment   
3  Second-Hand Dwelling house /Apartment   
4  Second-Hand Dwelling house /Apartment   

                           Property Size Description  
0                                                NaN  
1  greater than or equal to 38 sq metres and less...  
2                                                NaN  
3                                                NaN  
4                                                NaN

#### 4.2  Check for the DataFrame dimensionality with pandas .info() method

The .info() method in Pandas provides valuable insights about the DataFrame. The information contains the number of columns, column labels, column data types, memory usage, range index, and the number of cells in each column (non-null values). The info() method does not return any value, it prints the information. [[1]](https://www.w3schools.com/python/pandas/ref_df_info.asp) 

The output of the .info() method consists of several key components: [[2]](https://machinelearningtutorials.org/a-comprehensive-guide-to-using-the-pandas-dataframe-info-method/)

- The total number of rows (entries) in the DataFrame.

- A summary of each column, including:
  - The column name
  - The number of non-null values
  - The data type of the column
  - The memory usage of the column

The dataset contains the following columns:

- Date of Sale (dd/mm/yyyy): Date when the property was sold.

- Address: Address of the property.

- County: The county in which the property is located.

- Eircode: Eircode of the property.

- Price in Euro: The sale price of the property in Euros.

- Not Full Market Price: Indicates whether the sale was at the full market price.

- VAT Exclusive: Indicates whether VAT was excluded in the sale price.

- Description of Property: Type of property (e.g., second-hand or new).

- Property Size Description: Description of the property size (e.g., "greater than 125 sq metres").


In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 694503 entries, 0 to 694502
Data columns (total 9 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Date of Sale (dd/mm/yyyy)  694503 non-null  object 
 1   Address                    694503 non-null  object 
 2   County                     694503 non-null  object 
 3   Eircode                    166138 non-null  object 
 4   Price in Euro              694503 non-null  float64
 5   Not Full Market Price      694503 non-null  object 
 6   VAT Exclusive              694503 non-null  object 
 7   Description of Property    694503 non-null  object 
 8   Property Size Description  52830 non-null   object 
dtypes: float64(1), object(8)
memory usage: 47.7+ MB


Upon review, I can gather the following information:

- The DataFrame contains 694503 rows and 9 columns.
- The columns are: Date of Sale", ""Address", "County", "Eircode", "Price in Euro", "Not Full Market Price","VAT Exclusive", "Description of Property" and "Property Size Description".
- Some columns have non-null values, indicating that there are missing values in this dataset. Two variables of the columns have missing values.
- One quantitative variables is numeric with type float64: "Price in Euro". 
- Eight qualitative variables are categorical with type object: "Date of Sale", ""Address", "County", "Eircode", "Not Full Market Price","VAT Exclusive", "Description of Property" and "Property Size Description".
- The memory usage of this DataFrame is approximately 47.7 MB.

###### [1] [w3schools Pandas DataFrame info() Method](https://www.w3schools.com/python/pandas/ref_df_info.asp)
###### [2] [Understanding the .info output - Machine Learning Tutorials](https://machinelearningtutorials.org/a-comprehensive-guide-to-using-the-pandas-dataframe-info-method/)

#### 4.3  Generate descriptive statistics with pandas df.describe method

This analysis provides generalized descriptive statistics that summarises the central tendency of the data, the dispersion, and the shape of the dataset’s distribution. It also provides helpful information on missing NaN data. It includes the following statistics: [[3]](https://www.pythonlore.com/exploring-pandas-dataframe-describe-for-descriptive-statistics/)  [[4]](https://pandas.pydata.org/pandas-docs/version/0.20.2/generated/pandas.DataFrame.describe.html)

By default, the describe() function only generates descriptive statistics for numeric columns in a pandas DataFrame. I specifying include='all' which will force pandas to generate summaries for all columns in the dataframe. Some data types don’t have any information. Pandas marks them as NaN.


- Count: This represents the number of non-null (non-empty) values in the dataset for each column.

- Unique: This will show the number of unique values in the column

- Top: Ths will display the most common value in the column

- Frequency: This will show the frequency of the top value within the column.

- Mean: This will display the average value for each column.

- Standard deviation: It indicates how spread out the values are around the mean. A higher standard deviation means the values are more spread out from the mean, while a lower standard deviation means the values are closer to the mean.

- Minimum: It represents the lowest value in each column.

- The default percentiles of the describe function are 25th, 50th, and 75th percentile or (0.25, 0.5, and 0.75).
  
- First quartile (25th percentile): 25% of the data values are below this value.

- Second quartile (50th percentile): It represents the median, the middle value of the dataset.

- Third quartile (75th percentile): 75% of the data values are below this value.

- Maximum: It represents the highest value in the dataset.

In [68]:
df.describe(include="all")


Date of Sale (dd/mm/yyyy)               Address  County  Eircode  \
count                     694503                694503  694503   166138   
unique                      5145                622025      26   160850   
top                   22/12/2014  Broomfield, Midleton  Dublin  D24W9NN   
freq                        1542                    21  217653       34   
mean                         NaN                   NaN     NaN      NaN   
std                          NaN                   NaN     NaN      NaN   
min                          NaN                   NaN     NaN      NaN   
25%                          NaN                   NaN     NaN      NaN   
50%                          NaN                   NaN     NaN      NaN   
75%                          NaN                   NaN     NaN      NaN   
max                          NaN                   NaN     NaN      NaN   

        Price in Euro Not Full Market Price VAT Exclusive  \
count    6.945030e+05                694503        694503   
unique            NaN                     2             2   
top               NaN                    No            No   
freq              NaN                659594        579152   
mean     2.977859e+05                   NaN           NaN   
std      1.024681e+06                   NaN           NaN   
min      5.001000e+03                   NaN           NaN   
25%      1.350000e+05                   NaN           NaN   
50%      2.270000e+05                   NaN           NaN   
75%      3.400000e+05                   NaN           NaN   
max      2.250000e+08                   NaN           NaN   

                      Description of Property  \
count                                  694503   
unique                                      5   
top     Second-Hand Dwelling house /Apartment   
freq                                   576951   
mean                                      NaN   
std                                       NaN   
min                                       NaN   
25%                                       NaN   
50%                                       NaN   
75%                                       NaN   
max                                       NaN   

                                Property Size Description  
count                                               52830  
unique                                                  6  
top     greater than or equal to 38 sq metres and less...  
freq                                                38096  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

Key interpretations can be made from this function, both for the quantitative and qualitative data. 

Null values: Confirming the observations made from .info, the count of THE "Eircode" and "Property Size Description" rows does not match the count of values the other columns, indicating missing data.

###### [3] [Understanding the Output of pandas.DataFrame.describe](https://www.pythonlore.com/exploring-pandas-dataframe-describe-for-descriptive-statistics/)
###### [4] [Pandas Documentation on pandas.DataFrame.describe](https://pandas.pydata.org/pandas-docs/version/0.20.2/generated/pandas.DataFrame.describe.html)

#### 4.4 Check for missing data using df.isna

In [69]:
print(df.isna().sum())

Date of Sale (dd/mm/yyyy)         0
Address                           0
County                            0
Eircode                      528365
Price in Euro                     0
Not Full Market Price             0
VAT Exclusive                     0
Description of Property           0
Property Size Description    641673
dtype: int64


Two of the colums, "Eircode" and "Property Size Description" have missing values. I will keep this in  mind throughout the project.

## 5.   Data Analysis

***

### 5.1  Analysis of Property Prices for all years in Ireland (2010-2024)


Before the analysis, I need to set to datetime. 

In [70]:
df["Date of Sale (dd/mm/yyyy)"]

0         01/01/2010
1         03/01/2010
2         04/01/2010
3         04/01/2010
4         04/01/2010
             ...    
694498    22/11/2024
694499    22/11/2024
694500    22/11/2024
694501    22/11/2024
694502    22/11/2024
Name: Date of Sale (dd/mm/yyyy), Length: 694503, dtype: object

In [71]:
df["Date of Sale"] = pd.to_datetime(df["Date of Sale (dd/mm/yyyy)"], format="%d/%m/%Y")

In [72]:
df['Date of Sale']

0        2010-01-01
1        2010-01-03
2        2010-01-04
3        2010-01-04
4        2010-01-04
            ...    
694498   2024-11-22
694499   2024-11-22
694500   2024-11-22
694501   2024-11-22
694502   2024-11-22
Name: Date of Sale, Length: 694503, dtype: datetime64[ns]

In [73]:
df["Year"] = df["Date of Sale"].dt.year
print(df["Year"])

0         2010
1         2010
2         2010
3         2010
4         2010
          ... 
694498    2024
694499    2024
694500    2024
694501    2024
694502    2024
Name: Year, Length: 694503, dtype: int32


In [74]:
df["Month"] = df["Date of Sale"].dt.month
print (df["Month"])

0          1
1          1
2          1
3          1
4          1
          ..
694498    11
694499    11
694500    11
694501    11
694502    11
Name: Month, Length: 694503, dtype: int32


#### 5.1.1 Average property prices by county (2010-2024)

In [75]:
avg_price_by_county = df.groupby('County')['Price in Euro'].mean().sort_values(ascending=False).round()
print(avg_price_by_county)

County
Dublin       467700.0
Wicklow      372285.0
Kildare      310133.0
Meath        277372.0
Cork         259692.0
Galway       236303.0
Kilkenny     220146.0
Louth        216022.0
Limerick     202135.0
Wexford      195099.0
Laois        191823.0
Waterford    191452.0
Kerry        188500.0
Clare        188241.0
Westmeath    185704.0
Carlow       181083.0
Offaly       169068.0
Tipperary    161583.0
Sligo        157058.0
Monaghan     156578.0
Mayo         150320.0
Cavan        144464.0
Donegal      137105.0
Roscommon    127897.0
Longford     122406.0
Leitrim      122075.0
Name: Price in Euro, dtype: float64


#### 5.1.2 Average property prices by month (2010-2024)

In [76]:
avg_price_by_month = df.groupby('Month')['Price in Euro'].mean().round()
print(avg_price_by_month)

Month
1     283823.0
2     293714.0
3     287168.0
4     288848.0
5     282617.0
6     303963.0
7     306255.0
8     304590.0
9     307859.0
10    306818.0
11    295199.0
12    300688.0
Name: Price in Euro, dtype: float64


#### 5.1.3 Number of occurrences of each property type

In [77]:
description_of_property_counts = df["Description of Property"].value_counts()
print(description_of_property_counts)

Description of Property
Second-Hand Dwelling house /Apartment    576951
New Dwelling house /Apartment            117508
Teach/arasan Conaithe Athaimhe               40
Teach/arasan Conaithe Nua                     3
Teach/?ras?n C?naithe Nua                     1
Name: count, dtype: int64


#### 5.1.4 Full Market price comparision

In [79]:
full_market_price_counts = df["Not Full Market Price"].value_counts()
print(full_market_price_counts)

Not Full Market Price
No     659594
Yes     34909
Name: count, dtype: int64


#### 5.1.5 Number of property sales by county

In [80]:
county_counts = df["County"].value_counts()
print(county_counts)

County
Dublin       217653
Cork          76338
Kildare       37418
Galway        33874
Meath         28622
Limerick      25602
Wexford       24193
Wicklow       23105
Kerry         19460
Waterford     19128
Louth         18966
Donegal       18693
Tipperary     18684
Mayo          16667
Clare         15650
Westmeath     13446
Laois         11194
Kilkenny      11024
Cavan         10466
Sligo         10208
Roscommon      9976
Offaly         8806
Carlow         7677
Leitrim        6224
Longford       5954
Monaghan       5475
Name: count, dtype: int64


#### 5.1.6 County with the highest and lowest average property prices

In [94]:
highest_avg_price_county = avg_price_by_county.idxmax()
highest_avg_price = int(avg_price_by_county.max())
#print(highest_avg_price_county)
#print(int(highest_avg_price))
print(f"{highest_avg_price_county} has the highest average property price in Ireland, with an average price of €{highest_avg_price}.")

Dublin has the highest average property price in Ireland, with an average price of €467700.


In [95]:
lowest_avg_price_county = avg_price_by_county.idxmin()
lowest_avg_price = int(avg_price_by_county.min())
#print(lowest_avg_price_county)
#print(lowest_avg_price)
print(f"{lowest_avg_price_county} has the lowest average property price in Ireland, with an average price of €{lowest_avg_price}.")

Leitrim has the lowest average property price in Ireland, with an average price of €122075.


### 5.2  Analysis

5.2.1

5.2.2

5.2.3

## 6.   References

***

***
## End

